client/session.go

```diff
@@ -1437,20 +1437,27 @@ func (s *session) streamBlocksBatchFromPeer(
 			continue
 		}
 
+		missed := 0
 		for j := range result.Elements[i].Blocks {
 			if j >= len(batch[i].blocks) {
 				s.log.Errorf("stream blocks response from peer %s returned more blocks than expected", peer.Host().String())
 				break
 			}
 
-			block := result.Elements[i].Blocks[j]
+			// Index of the received block could be offset by missed blocks
+			block := result.Elements[i].Blocks[j-missed]
+
 			if block.Start != batch[i].blocks[j].start.UnixNano() {
+				missed++
```

Comment:
Does this potentially introduce a cascading effect of errors? Ie, one start time mismatched, but all the others matches, then n would be compared with n-1 for the rest of the blocks, causing them all to be wrong. Why not use set overlap and disjoint sets to determine what to reattempt?